ci/ray_ci/automation/docker_tags_lib.py

```diff
@@ -0,0 +1,259 @@
+class DockerHubRateLimitException(Exception):
+    """
+    Exception for Docker Hub rate limit exceeded.
+    """
+
+    def __init__(self):
+        super().__init__("429: Rate limit exceeded for Docker Hub.")
+
+
+class RetrieveImageConfigException(Exception):
+    """
+    Exception for failing to retrieve image config.
+    """
+
+    def __init__(self):
+        super().__init__("Failed to retrieve image config.")
+
+
+class AuthTokenException(Exception):
+    """
+    Exception for failing to retrieve auth token.
+    """
+
+    def __init__(self, message: str):
+        super().__init__(f"Failed to retrieve auth token from {message}.")
+
+
+def get_docker_auth_token(namespace: str, repository: str) -> Optional[str]:
+    """
+    Retrieve Docker Registry token.
+
+    Args:
+        namespace: Docker namespace
+        repository: Docker repository
+
+    Returns:
+        Auth token for Docker.
+    """
+    service, scope = (
+        "registry.docker.io",
+        f"repository:{namespace}/{repository}:pull",
+    )
+    auth_url = f"https://auth.docker.io/token?service={service}&scope={scope}"
+    response = requests.get(auth_url)
+    if response.status_code != 200:
+        raise AuthTokenException(f"Docker. Error code: {response.status_code}")
+    token = response.json().get("token", None)
+    return token
+
+
+def get_docker_hub_auth_token():
+    """
+    Retrieve Docker Hub auth token.
+    """
+    username = os.environ["DOCKER_HUB_USERNAME"]
+    password = os.environ["DOCKER_HUB_PASSWORD"]
+
+    url = "https://hub.docker.com/v2/users/login"
+    json_body = {
+        "username": username,
+        "password": password,
+    }
+    headers = {"Content-Type": "application/json"}
+    response = requests.post(url, headers=headers, json=json_body)
+    if response.status_code != 200:
+        raise AuthTokenException(f"Docker Hub. Error code: {response.status_code}")
+    return response.json().get("token", None)
+
+
+def _get_git_log(n_days: int = 30):
+    return subprocess.check_output(
+        [
+            "git",
+            "log",
+            f"--until='{n_days} days ago'",
+            "--pretty=format:%H",
+        ],
+        text=True,
+    )
+
+
+def list_recent_commit_short_shas(
+    n_days: int = 30,
+) -> List[str]:
+    """
+    Get list of recent commit SHAs (short version, first 6 char) on ray master branch.
+
+    Args:
+        n_days: Number of days to go back in git log.
+
+    Returns:
+        List of recent commit SHAs (6 char).
+    """
+    commit_shas = _get_git_log(n_days=n_days)
+    short_commit_shas = [
+        commit_sha[:6] for commit_sha in commit_shas.split("\n") if commit_sha
+    ]
+    return short_commit_shas
+
+
+def _call_crane_config(tag: str):
+    try:
+        return subprocess.check_output(
+            ["crane", "config", tag],
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+    except subprocess.CalledProcessError as e:
+        return e.output
```

Comment:
maybe just call the OCI API to get the image config?

Reply:
done